src/CommonProviderImplementation/AssemblyReplacer.fs

```diff
@@ -14,145 +15,128 @@ open Microsoft.FSharp.Quotations.Patterns
+      | true, newT -> newT
+      | false, _ -> 
+            let asms = (if fwd then referencedAssemblies else designTimeAssemblies)
+            let fullName = fixName t.FullName
 
-  let private replace asmMappings (original, originalAsms) f =
-    replaceLazy asmMappings (lazy original, originalAsms) f
+            // For normal type provider hosts like fsc.exe System.Void is a special case and these
+            // hosts expect it to always be provided as [FSharp.Core 4.4.0.0]typeof<System.Void>.  
+            // This is really a mistake in ExtensionTyping.fs in the F# compiler which calls 
+            // typeof<System.Void>.Equals(ty).
+            if fullName = "System.Void" then typeof<System.Void> else
 
-  let rec private getType asmCandidates (t:Type) asmMappings ((typeCache:Dictionary<_,_>, _) as caches) =
-    let getType (asm:Assembly) =
-      match typeCache.TryGetValue((asm, t)) with
-      | true, t -> Some t
-      | false, _ ->
-          let fixName (fullName:string) =
-            if fullName.StartsWith("FSI_") then 
-              // when F# Interactive is the host of the design time assembly,
-              // all namespaces are prefixed with FSI_, in the runtime assembly
-              // the name won't have that prefix
-              fullName.Substring(fullName.IndexOf('.') + 1)
-            else 
-              fullName
-          let getType fullName =
-            if asm.FullName.StartsWith "FSI-ASSEMBLY" then
-              // when F# Interactive is the host of the design time assembly,
-              // for each type in the runtime assembly there might be multiple
-              // versions (FSI_0001.FullTypeName, FSI_0002.FullTypeName, etc).
-              // Get the last one.
-              asm.GetTypes() 
-              |> Seq.filter (fun t -> (fixName t.FullName) = fullName)
-              |> Seq.sortBy (fun t -> t.FullName)
-              |> Seq.last
-            else
-              asm.GetType fullName
-          let newT =
-            if t.IsGenericType && not t.IsGenericTypeDefinition then 
-              let genericType = t.GetGenericTypeDefinition()
-              let newT = getType (fixName genericType.FullName)
-              if newT = null then 
-                null
-              else
-                let typeArguments = 
-                  t.GetGenericArguments()
-                  |> Array.map (replaceType asmMappings caches)
-                newT.MakeGenericType(typeArguments)
-            else 
-              getType (fixName t.FullName)
-          if newT = null then
-            None
-          else
-            typeCache.Add((asm, t), newT)
-            Some newT
-    match List.choose getType asmCandidates with
-    | [ t ] -> t
-    | _ -> failwithf "Type '%O' not found in %A" t (asmCandidates |> List.map (fun asm -> asm.Location))
+            match Array.choose (tryGetTypeFromAssembly fullName) asms |> Seq.distinct |> Seq.toArray with
+            //| [| (newT, canSave) |] -> 
```

Comment:
Humm, could it happen that FSharp.Core 4.3.0.0 is referenced because that's what FSharp.Data is dependent on and FSharp.Core 4.3.1.0 or 4.4.0.0 is also referenced because that's the version currently loaded in the TP host inside the compiler? I think I've seen something like that before